pantry/src/pantry.rs

```diff
@@ -0,0 +1,431 @@
+
+        // XXX activation number going away?
+        volume.activate(0).await?;
+
+        info!(self.log, "volume {} activated ok", volume_id);
+
+        entries.insert(
+            volume_id.clone(),
+            Arc::new(Mutex::new(PantryEntry {
+                volume,
+                volume_construction_request,
+            })),
+        );
+
+        info!(self.log, "volume {} constructed and inserted ok", volume_id);
+
+        Ok(())
+    }
+
+    pub async fn entry(
+        &self,
+        volume_id: String,
+    ) -> Result<Arc<Mutex<PantryEntry>>, HttpError> {
+        let entries = self.entries.lock().await;
+        match entries.get(&volume_id) {
+            Some(entry) => {
+                let entry = entry.clone();
+                drop(entries);
+                Ok(entry)
+            }
+
+            None => {
+                error!(self.log, "volume {} not in pantry", volume_id);
+
+                Err(HttpError::for_not_found(None, volume_id))
+            }
+        }
+    }
+
+    pub async fn is_job_finished(
+        &self,
+        job_id: String,
+    ) -> Result<bool, HttpError> {
+        let jobs = self.jobs.lock().await;
+        match jobs.get(&job_id) {
+            Some(join_handle) => Ok(join_handle.is_finished()),
+
+            None => {
+                error!(self.log, "job {} not a pantry job", job_id);
+
+                Err(HttpError::for_not_found(None, job_id.to_string()))
+            }
+        }
+    }
+
+    pub async fn get_job_result(
+        &self,
+        job_id: String,
+    ) -> Result<Result<()>, HttpError> {
+        let mut jobs = self.jobs.lock().await;
+
+        // Remove the job from the list of jobs, then await on the join handle.
+        // If this errors, then the job has failed in some way, so don't leave
+        // it in the list of jobs.
+        match jobs.remove(&job_id) {
+            Some(join_handle) => {
+                let result = join_handle.await.map_err(|e| {
+                    HttpError::for_internal_error(e.to_string())
+                })?;
+                jobs.remove(&job_id);
+                Ok(result)
+            }
+
+            None => {
+                error!(self.log, "job {} not a pantry job", job_id);
+
+                Err(HttpError::for_not_found(None, job_id.to_string()))
+            }
+        }
+    }
+
+    pub async fn import_from_url(
+        &self,
+        volume_id: String,
+        url: String,
+        expected_digest: Option<ExpectedDigest>,
+    ) -> Result<String, HttpError> {
+        let entry = self.entry(volume_id).await?;
+        let entry = entry.clone();
+
+        let join_handle = tokio::spawn(async move {
+            entry
+                .lock()
+                .await
+                .import_from_url(url, expected_digest)
+                .await
+        });
+
+        let mut jobs = self.jobs.lock().await;
+        let job_id = Uuid::new_v4().to_string();
```

Comment:
Out of curiosity, why a UUID here?  Does the caller (Nexus) make use of that UUID to track/follow jobs?

Reply:
exactly - just a unique job identifier